### ISSUE TITLE
[Bug] : min()/max()/clamp() get merged over their fallback values

### ISSUE THREAD

**Issue by wongjn** [cssnano/cssnano#897]
**Describe the bug**
Properties merged by `postcss-merge-longhand` package will merge `min()`/`max()`/`clamp()` over their fallback values.

**To Reproduce**
Before (padding is not the only property to suffer with this):
```css
h1 {
  padding: 1em;
  padding: min(1em, 1vw);
}
```
After:
```css
h1 {
  padding: min(1em, 1vw);
}
```

**Expected behavior**
After:
```css
h1 {
  padding: 1em;
  padding: min(1em, 1vw);
}
```

**Additional context**

I think it is important to keep the fallback values as browser support is a little more sparse than custom properties, where the package would preserve in the above situations.

**Comment by anikethsaha** [comment]
Currently, merge longhand plugin doesn't check for compatibility and it merges specifically longhand props. So it's the intended behavior. 
But checking compatibility can be enhancement before evening traversing that node.

**Comment by ludofischer** [comment]
Confirmed in 5.0.0-rc.2

**Comment by ludofischer** [comment]
Possibly duplicate of https://github.com/cssnano/cssnano/issues/846

**Comment by alexander-akait** [comment]
@ludofischer Maybe the best solution here - check if values have the same type and remove duplication only if type is same, i.e. if we have `10px` on top and `20px` on bottom, we can safely remove top

**Comment by ludofischer** [comment]
> @ludofischer Maybe the best solution here - check if values have the same type and remove duplication only if type is same, i.e. if we have `10px` on top and `20px` on bottom, we can safely remove top

I was thinking about a rougher solution, not merging if two identical properties come one after the other in the original CSS, as that is an indication that one is meant as a fallback.
How do you identify the same 'type'? I'm afraid it would turn into a lot of code and still be very error prone. For example, in #846 they complain about these two getting merged:
```
.cls {
padding-top: calc(1rem + constant(safe-area-inset-top));
}
.cls {
padding-top: calc(1rem + env(safe-area-inset-top));
}
```
In this case, both are `calc()` expressions, to recognise that they are a different type we would need to analyse the contents of the calc expression,  recognize individual features (functions and special values like `safe-area-inset`) and preserve if the browser support is not the same?

**Comment by alexander-akait** [comment]
Yep, it is required more better parser... As workaround we can't disable merging for calc/env/constant/var (maybe for all function?)

**Comment by ludofischer** [comment]
> Yep, it is required more better parser... As workaround we can't disable merging for calc/env/constant/var (maybe for all function?)

I think that's the way to go. Merging is disabled already when variables, the only question is that function calls are harder to detect, I think we really need to parse the value before deciding to merge.

**Comment by ludofischer** [comment]
Unfortunately it does not look that simple. If we skip merging every time there is a function we're going to skip merging everything that contains a color like `rgba()`. To detect a fallback, we would need to find the case where the same exact property is declared twice in a row, but the `merge-longhand` plugin always runs `explode` before the merge, that means in the merge phase the original CSS is already lost.  I'm afraid it's quite hard to make the merge longhand transform really 'safe', at least with the way the `postcss-merge-longhand` plugin is written now.

**Comment by maranomynet** [comment]
> I was thinking about a rougher solution, not merging if two identical properties come one after the other in the original CSS, as that is an indication that one is meant as a fallback.

I think this is a very sensible approach. :+1:

**Comment by maranomynet** [comment]
For the time being, I'm working around this in SCSS with this mixin:

```scss
// Hack to generate an extra selector block to stop cssnano
// from destructively merging adjacent properties – e.g. when
// providing a fallback value for older browsers.
//
// Remove as soon as this issue has been resolved:
// https://github.com/cssnano/cssnano/issues/897
@mixin avoidCssnanoMerging() {
  & {
    @content;
  }
}
```

Usage example:

```scss
h1 {
  padding: 1em;
  @include avoidCssnanoMerging(){
    padding: min(1em, 1vw);
  }
}
```

Looking forward to some proper fix.

**Comment by ludofischer** [comment]
> Looking forward to some proper fix.

Last time I've looked it is not that easy, because the plugin destroys the original AST before attempting the merging.

**Comment by ludofischer** [comment]
This has been fixed on the latest release
